Error response for merchant center link/create operation.

ankitrox opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes: https://linear.app/a8c/issue/GOOWOO-339/merchant-center-account-creationlinking-fails-insert-link

_Replace this with a good description of your changes & reasoning._


### Screenshots:

<img width="686" height="580" alt="Screenshot 2025-12-12 at 10 57 12 PM" src="https://github.com/user-attachments/assets/83679087-f10b-4877-8311-0de8974f7747" />

### Detailed test instructions

You may need to checkout to `test/GOOWOO-214-integration` [branch](https://github.com/woocommerce/google-listings-and-ads/pull/3160) to test these changes on FE. This feature branch is merged in the integration branch.

1. Make sure test proxy is running.

2. Go to step 1 of the onboarding, make sure Jetpack and Google accounts are connected and Ads and MC cards are visible, but are not connected.

3. Try connecting Ads and MC accounts one by one, the error would be displayed in corresponding cards as shown in screenshot.

**Note:** We also have corresponding E2E tests for these scenarios as well as for accounts auto-creation scenrio in integration branch given above.

### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

>

mukeshpanchal27 commented on src/API/Google/Middleware.php:
```diff
@@ -154,6 +155,18 @@ protected function create_merchant_account_request( string $name, string $site_u
+						}
+					}
+				}
+			}
```
This code is duplicated across the files. Please factor it out into a shared utility function.

mukeshpanchal27 commented on src/MerchantCenter/AccountService.php:
```diff
@@ -141,6 +141,22 @@ public function use_existing_account_id( int $account_id ): void {
+						'code'  => 'API_ERROR',
+						'error' => $error,
+					],
+					$e->getCode() ?: 400
+				);
+			}
```
Same as previous message

mukeshpanchal27 reviewed (commented):
Thanks for the PR. Could you please add testing instructions to the PR?

It would also be great if you include the Proxy testing: https://github.com/woocommerce/google-listings-and-ads/tree/develop/tests/proxy so we can test it easily.

ankitrox commented:
Thank you @mukeshpanchal27 for reviewing the PR and adding the feedback.

I have updated the proxy to accommodate the changes and also updated the test instructions to test the PR changes.

Over to you for another review.

mukeshpanchal27 commented:
When i test this with proxy i get below errors:

<img width="1565" height="794" alt="Screenshot 2025-12-18 at 9 47 38 AM" src="https://github.com/user-attachments/assets/4bf37ce1-c7f5-4953-babf-91eb7a4532fb" />

<img width="1717" height="846" alt="Screenshot 2025-12-18 at 9 48 29 AM" src="https://github.com/user-attachments/assets/0b68e7cc-7f4b-426d-976b-6323555bb9d5" />

ankitrox commented:
@mukeshpanchal27 Thanks for looking into it.
Can you please confirm if you are testing on the `test/GOOWOO-214-integration` branch?

mukeshpanchal27 commented on src/API/Google/Middleware.php:
```diff
@@ -264,11 +266,20 @@ public function claim_merchant_website( bool $overwrite = false ): bool {
+					}
+				}
+			}
+			throw new Exception( $message, $status, $e );
```
Why do we reference `$e` instead of `$errors`, considering `$errors` is derived from `$e`?

ankitrox replied:
We can skip adding the error object completely. `$errors` can't be added as third parameter needs to be of type `Throwable` and `$errors` is array.

mukeshpanchal27 commented on src/API/Site/Controllers/MerchantCenter/AccountController.php:
```diff
@@ -171,6 +172,22 @@ protected function setup_account_callback( string $action = 'setup_account' ): c
+					$data    = $e->get_response_data();
+					$status  = $e->getCode() ?: 400;
+					$error   = is_array( $data ) ? ( $data['error'] ?? [] ) : [];
+					$message = is_array( $error ) && isset( $error['message'] ) ? (string) $error['message'] : $e->getMessage();
```
This should be moved inside the if condition, since this variable only needs to be set when the error code is not `API_ERROR`.

ankitrox replied:
Thanks, nice observation. I updated it.

mukeshpanchal27 reviewed (commented):
Thanks for the update. It looks solid overall. I’ve left a few questions. Let me know if you need any clarification.

mukeshpanchal27 reviewed (approved):
Thanks @ankitrox for the updates. The changes look good in my testing.

Make sure to merge develop branch so unit tests get passed.

ankitrox committed:
Merge branch 'feature/GOOWOO-214-better-user-guidance' into add/GOOWOO-339-merchant-linking-error-response

ankitguptaindia commented:
@asvinb / @ankitrox, this branch has some conflicts. Could you please fix and merge this into the feature branch so I can test this with other features together?

asvinb committed:
Merge branch 'feature/GOOWOO-214-better-user-guidance' into add/GOOWOO-339-merchant-linking-error-response

asvinb committed:
Update bundlesize

ankitguptaindia commented:
QA Status: Approved ✅ 
---



### Error message when MC & Ad account has a connection error: 

<img width="1222" height="882" alt="Screenshot 2026-04-17 at 4 29 50 PM" src="https://github.com/user-attachments/assets/e4294890-2b61-4c77-a812-3987cde97922" />



### Connection Flow when no error with Ads & MC account: 

https://github.com/user-attachments/assets/7697e113-fba3-41b6-b1d8-f873c4487da7
